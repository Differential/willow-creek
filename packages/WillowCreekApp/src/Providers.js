import { nest } from 'recompose';
import { Providers } from '@apollosproject/ui-kit';
import ClientProvider from './client';

export default nest(ClientProvider, Providers);
