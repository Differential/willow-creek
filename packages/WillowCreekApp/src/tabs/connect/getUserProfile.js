import gql from 'graphql-tag';

export default gql`
  query {
    currentUser {
      id
      profile {
        firstName
        lastName
        location
        email
        nickName
        photo {
          uri
        }
      }
    }
  }
`;
