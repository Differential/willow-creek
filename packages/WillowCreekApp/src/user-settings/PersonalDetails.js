import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { SafeAreaView } from 'react-native';
import { Query, Mutation } from 'react-apollo';
import { Formik } from 'formik';
import * as Yup from 'yup';
import { KeyboardAwareScrollView } from 'react-native-keyboard-aware-scroll-view';

import {
  TextInput,
  PaddedView,
  FlexedView,
  Button,
  ButtonLink,
  TableView,
} from '@apollosproject/ui-kit';

import getUserProfile from '../tabs/connect/getUserProfile';
import updateCurrentUser from './updateCurrentUser';

class PersonalDetails extends PureComponent {
  static navigationOptions = ({ navigation }) => ({
    title: 'Personal Details',
    headerLeft: null,
    headerRight: (
      <PaddedView vertical={false}>
        <ButtonLink onPress={() => navigation.goBack()}>Cancel</ButtonLink>
      </PaddedView>
    ),
  });

  static propTypes = {
    navigation: PropTypes.shape({
      getParam: PropTypes.func,
      navigate: PropTypes.func,
      goBack: PropTypes.func,
    }),
  };

  renderForm = (props) => (
    <FlexedView>
      <KeyboardAwareScrollView>
        <TableView>
          <PaddedView>
            <TextInput
              label="First Name"
              type="text"
              value={props.values.firstName}
              error={props.touched.firstName && props.errors.firstName}
              onChangeText={(text) => props.setFieldValue('firstName', text)}
            />
            <TextInput
              label="Last Name"
              type="text"
              value={props.values.lastName}
              error={props.touched.lastName && props.errors.lastName}
              onChangeText={(text) => props.setFieldValue('lastName', text)}
            />
          </PaddedView>
        </TableView>
        <TableView>
          <PaddedView>
            <TextInput
              label="Email"
              type="email"
              value={props.values.email}
              error={props.touched.email && props.errors.email}
              onChangeText={(text) => props.setFieldValue('email', text)}
            />
          </PaddedView>
        </TableView>
      </KeyboardAwareScrollView>
      <SafeAreaView>
        <PaddedView>
          <Button
            disabled={!props.isValid || props.isSubmitting}
            onPress={props.handleSubmit}
            title="Save"
            loading={props.isSubmitting}
          />
        </PaddedView>
      </SafeAreaView>
    </FlexedView>
  );

  render() {
    return (
      <Query query={getUserProfile} fetchPolicy="cache-and-network">
        {({ data: { currentUser = { profile: {} } } = {} }) => {
          const { firstName, lastName, email } = currentUser.profile;

          return (
            <Mutation
              mutation={updateCurrentUser}
              update={async (cache, { data: { updateProfileFields } }) => {
                await cache.writeQuery({
                  query: getUserProfile,
                  data: {
                    currentUser: {
                      ...currentUser,
                      profile: {
                        ...currentUser.profile,
                        firstName: updateProfileFields.firstName,
                        lastName: updateProfileFields.lastName,
                        email: updateProfileFields.email,
                      },
                    },
                  },
                });
              }}
            >
              {(updateDetails) => (
                <Formik
                  initialValues={{ firstName, lastName, email }}
                  validationSchema={Yup.object().shape({
                    firstName: Yup.string().required('First Name is required!'),
                    lastName: Yup.string().required('Last Name is required!'),
                    email: Yup.string()
                      .email('Invalid email address')
                      .required('Email is required!'),
                  })}
                  onSubmit={async (
                    variables,
                    { setSubmitting, setFieldError }
                  ) => {
                    try {
                      await updateDetails({ variables });
                      await this.props.navigation.goBack();
                    } catch (e) {
                      const { graphQLErrors } = e;
                      if (
                        graphQLErrors.length &&
                        graphQLErrors.find(({ message }) =>
                          message.includes('User already exists')
                        )
                      ) {
                        setFieldError(
                          'email',
                          'There is already a user with this email'
                        );
                      } else {
                        setFieldError(
                          'email',
                          'Unknown error. Please try again later.'
                        );
                      }
                    }
                    setSubmitting(false);
                  }}
                >
                  {this.renderForm}
                </Formik>
              )}
            </Mutation>
          );
        }}
      </Query>
    );
  }
}

export default PersonalDetails;
