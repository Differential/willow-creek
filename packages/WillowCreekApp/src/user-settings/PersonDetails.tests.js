import React from 'react';
import renderer from 'react-test-renderer';
import wait from 'waait';

import Providers from 'WillowCreekApp/src/Providers';

import getUserProfile from '../tabs/connect/getUserProfile';
import PersonalDetails from './PersonalDetails';

describe('PersonalDetails component', () => {
  it('renders PersonalDetails when logged in', async () => {
    const mock = {
      request: {
        query: getUserProfile,
      },
      result: {
        data: {
          currentUser: {
            profile: {
              firstName: 'Isaac',
              lastName: 'Hardy',
              nickName: 'Ike',
              email: 'isaac.hardy@newspring.cc',
            },
          },
        },
      },
    };
    const navigation = {
      navigate: jest.fn(),
      getParam: jest.fn(),
      goBack: jest.fn(),
    };
    const tree = renderer.create(
      <Providers mocks={[mock]}>
        <PersonalDetails navigation={navigation} />
      </Providers>
    );
    await wait(0); // wait for response from graphql
    expect(tree).toMatchSnapshot();
  });
});
