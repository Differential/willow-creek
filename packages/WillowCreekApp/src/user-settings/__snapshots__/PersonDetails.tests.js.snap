// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PersonalDetails component renders PersonalDetails when logged in 1`] = `
<RCTSafeAreaView
  onLayout={[Function]}
  style={
    Object {
      "bottom": 0,
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
    }
  }
>
  <View
    onLayout={[Function]}
    style={
      Object {
        "flex": 1,
      }
    }
  />
  <View
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <RCTScrollView
        automaticallyAdjustContentInsets={false}
        contentInset={
          Object {
            "bottom": 0,
          }
        }
        enableAutomaticScroll={true}
        enableOnAndroid={false}
        enableResetScrollToCoords={true}
        extraHeight={75}
        extraScrollHeight={0}
        getScrollResponder={[Function]}
        handleOnScroll={[Function]}
        keyboardDismissMode="interactive"
        keyboardOpeningTime={250}
        keyboardSpace={0}
        onScroll={[Function]}
        resetKeyboardSpace={[Function]}
        scrollEventThrottle={1}
        scrollForExtraHeightOnAndroid={[Function]}
        scrollIntoView={[Function]}
        scrollToEnd={[Function]}
        scrollToFocusedInput={[Function]}
        scrollToPosition={[Function]}
        showsVerticalScrollIndicator={true}
        update={[Function]}
        viewIsInsideTabBar={false}
      >
        <View>
          <View
            style={
              Object {
                "backgroundColor": "#FFFFFF",
                "borderBottomWidth": 0.5,
                "borderColor": "rgba(165, 165, 165, 0.30000000000000004)",
                "borderLeftWidth": 0,
                "borderRightWidth": 0,
                "borderTopWidth": 0.5,
                "marginBottom": 16,
              }
            }
          >
            <View
              style={
                Object {
                  "paddingHorizontal": 16,
                  "paddingVertical": 16,
                }
              }
            >
              <View
                disabled={false}
                style={
                  Object {
                    "marginVertical": 16,
                  }
                }
              >
                <View>
                  <View
                    style={
                      Object {
                        "flexDirection": "row",
                        "justifyContent": "center",
                      }
                    }
                  >
                    <View
                      style={
                        Object {
                          "alignItems": "center",
                          "justifyContent": "center",
                        }
                      }
                    />
                    <View
                      style={
                        Object {
                          "flex": 1,
                          "opacity": 0,
                        }
                      }
                    >
                      <TextInput
                        allowFontScaling={true}
                        editable={true}
                        onBlur={[Function]}
                        onChangeText={[Function]}
                        onFocus={[Function]}
                        placeholderTextColor="rgba(48, 48, 48, 0.6)"
                        returnKeyType="done"
                        style={
                          Object {
                            "color": "#303030",
                            "fontFamily": "InterUI-Bold",
                            "fontSize": 14,
                            "height": 38,
                            "paddingVertical": 0,
                          }
                        }
                        type="text"
                        underlineColorAndroid="transparent"
                        value=""
                      />
                    </View>
                    <View
                      style={
                        Object {
                          "alignItems": "center",
                          "justifyContent": "center",
                        }
                      }
                    />
                  </View>
                  <View
                    onLayout={[Function]}
                    pointerEvents="none"
                    style={
                      Object {
                        "bottom": 0,
                        "justifyContent": "center",
                        "opacity": 1,
                        "position": "absolute",
                        "top": 0,
                        "transform": Array [
                          Object {
                            "scale": 1,
                          },
                          Object {
                            "translateX": 0,
                          },
                          Object {
                            "translateY": 0,
                          },
                        ],
                      }
                    }
                  >
                    <Text
                      style={
                        Object {
                          "backgroundColor": "transparent",
                          "color": "#505050",
                          "fontFamily": "InterUI-Bold",
                          "fontSize": 12,
                          "lineHeight": 17.28,
                          "paddingVertical": 4,
                        }
                      }
                    >
                      First Name
                    </Text>
                  </View>
                  <View
                    style={
                      Object {
                        "bottom": 0,
                        "left": 0,
                        "right": 0,
                      }
                    }
                  >
                    <View
                      style={
                        Object {
                          "backgroundColor": "#A5A5A5",
                          "bottom": -0.5,
                          "height": 0.5,
                        }
                      }
                    />
                    <View
                      style={
                        Object {
                          "bottom": 0,
                          "left": 0,
                          "right": 0,
                          "transform": Array [
                            Object {
                              "scaleX": 0.0001,
                            },
                          ],
                        }
                      }
                    >
                      <View
                        hasError={false}
                        style={
                          Object {
                            "backgroundColor": "#00676D",
                            "bottom": 0,
                            "height": 0.5,
                          }
                        }
                      />
                    </View>
                  </View>
                </View>
              </View>
              <View
                disabled={false}
                style={
                  Object {
                    "marginVertical": 16,
                  }
                }
              >
                <View>
                  <View
                    style={
                      Object {
                        "flexDirection": "row",
                        "justifyContent": "center",
                      }
                    }
                  >
                    <View
                      style={
                        Object {
                          "alignItems": "center",
                          "justifyContent": "center",
                        }
                      }
                    />
                    <View
                      style={
                        Object {
                          "flex": 1,
                          "opacity": 0,
                        }
                      }
                    >
                      <TextInput
                        allowFontScaling={true}
                        editable={true}
                        onBlur={[Function]}
                        onChangeText={[Function]}
                        onFocus={[Function]}
                        placeholderTextColor="rgba(48, 48, 48, 0.6)"
                        returnKeyType="done"
                        style={
                          Object {
                            "color": "#303030",
                            "fontFamily": "InterUI-Bold",
                            "fontSize": 14,
                            "height": 38,
                            "paddingVertical": 0,
                          }
                        }
                        type="text"
                        underlineColorAndroid="transparent"
                        value=""
                      />
                    </View>
                    <View
                      style={
                        Object {
                          "alignItems": "center",
                          "justifyContent": "center",
                        }
                      }
                    />
                  </View>
                  <View
                    onLayout={[Function]}
                    pointerEvents="none"
                    style={
                      Object {
                        "bottom": 0,
                        "justifyContent": "center",
                        "opacity": 1,
                        "position": "absolute",
                        "top": 0,
                        "transform": Array [
                          Object {
                            "scale": 1,
                          },
                          Object {
                            "translateX": 0,
                          },
                          Object {
                            "translateY": 0,
                          },
                        ],
                      }
                    }
                  >
                    <Text
                      style={
                        Object {
                          "backgroundColor": "transparent",
                          "color": "#505050",
                          "fontFamily": "InterUI-Bold",
                          "fontSize": 12,
                          "lineHeight": 17.28,
                          "paddingVertical": 4,
                        }
                      }
                    >
                      Last Name
                    </Text>
                  </View>
                  <View
                    style={
                      Object {
                        "bottom": 0,
                        "left": 0,
                        "right": 0,
                      }
                    }
                  >
                    <View
                      style={
                        Object {
                          "backgroundColor": "#A5A5A5",
                          "bottom": -0.5,
                          "height": 0.5,
                        }
                      }
                    />
                    <View
                      style={
                        Object {
                          "bottom": 0,
                          "left": 0,
                          "right": 0,
                          "transform": Array [
                            Object {
                              "scaleX": 0.0001,
                            },
                          ],
                        }
                      }
                    >
                      <View
                        hasError={false}
                        style={
                          Object {
                            "backgroundColor": "#00676D",
                            "bottom": 0,
                            "height": 0.5,
                          }
                        }
                      />
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
          <View
            style={
              Object {
                "backgroundColor": "#FFFFFF",
                "borderBottomWidth": 0.5,
                "borderColor": "rgba(165, 165, 165, 0.30000000000000004)",
                "borderLeftWidth": 0,
                "borderRightWidth": 0,
                "borderTopWidth": 0.5,
                "marginBottom": 16,
              }
            }
          >
            <View
              style={
                Object {
                  "paddingHorizontal": 16,
                  "paddingVertical": 16,
                }
              }
            >
              <View
                disabled={false}
                style={
                  Object {
                    "marginVertical": 16,
                  }
                }
              >
                <View>
                  <View
                    style={
                      Object {
                        "flexDirection": "row",
                        "justifyContent": "center",
                      }
                    }
                  >
                    <View
                      style={
                        Object {
                          "alignItems": "center",
                          "justifyContent": "center",
                        }
                      }
                    />
                    <View
                      style={
                        Object {
                          "flex": 1,
                          "opacity": 0,
                        }
                      }
                    >
                      <TextInput
                        allowFontScaling={true}
                        autoCapitalize="none"
                        autoCorrect={false}
                        editable={true}
                        keyboardType="email-address"
                        onBlur={[Function]}
                        onChangeText={[Function]}
                        onFocus={[Function]}
                        placeholderTextColor="rgba(48, 48, 48, 0.6)"
                        returnKeyType="done"
                        style={
                          Object {
                            "color": "#303030",
                            "fontFamily": "InterUI-Bold",
                            "fontSize": 14,
                            "height": 38,
                            "paddingVertical": 0,
                          }
                        }
                        type="email"
                        underlineColorAndroid="transparent"
                        value=""
                      />
                    </View>
                    <View
                      style={
                        Object {
                          "alignItems": "center",
                          "justifyContent": "center",
                        }
                      }
                    />
                  </View>
                  <View
                    onLayout={[Function]}
                    pointerEvents="none"
                    style={
                      Object {
                        "bottom": 0,
                        "justifyContent": "center",
                        "opacity": 1,
                        "position": "absolute",
                        "top": 0,
                        "transform": Array [
                          Object {
                            "scale": 1,
                          },
                          Object {
                            "translateX": 0,
                          },
                          Object {
                            "translateY": 0,
                          },
                        ],
                      }
                    }
                  >
                    <Text
                      style={
                        Object {
                          "backgroundColor": "transparent",
                          "color": "#505050",
                          "fontFamily": "InterUI-Bold",
                          "fontSize": 12,
                          "lineHeight": 17.28,
                          "paddingVertical": 4,
                        }
                      }
                    >
                      Email
                    </Text>
                  </View>
                  <View
                    style={
                      Object {
                        "bottom": 0,
                        "left": 0,
                        "right": 0,
                      }
                    }
                  >
                    <View
                      style={
                        Object {
                          "backgroundColor": "#A5A5A5",
                          "bottom": -0.5,
                          "height": 0.5,
                        }
                      }
                    />
                    <View
                      style={
                        Object {
                          "bottom": 0,
                          "left": 0,
                          "right": 0,
                          "transform": Array [
                            Object {
                              "scaleX": 0.0001,
                            },
                          ],
                        }
                      }
                    >
                      <View
                        hasError={false}
                        style={
                          Object {
                            "backgroundColor": "#00676D",
                            "bottom": 0,
                            "height": 0.5,
                          }
                        }
                      />
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </RCTScrollView>
      <RCTSafeAreaView>
        <View
          style={
            Object {
              "paddingHorizontal": 16,
              "paddingVertical": 16,
            }
          }
        >
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <View
              bordered={false}
              disabled={true}
              pill={true}
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#00676D",
                  "borderColor": "#00676D",
                  "borderRadius": 48,
                  "borderWidth": 2,
                  "elevation": 2,
                  "flexDirection": "row",
                  "height": 48,
                  "justifyContent": "center",
                  "opacity": 0.5,
                  "overflow": "hidden",
                  "paddingHorizontal": 16,
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "#FFFFFF",
                    "fontFamily": "InterUI-Bold",
                    "fontSize": 16,
                    "lineHeight": 23.04,
                  }
                }
              >
                Save
              </Text>
            </View>
          </View>
        </View>
      </RCTSafeAreaView>
    </View>
  </View>
</RCTSafeAreaView>
`;
